internal/pkg/service/buffer/storage/local/writer/disksync/disksync.go

```diff
@@ -0,0 +1,277 @@
+	}
+
+	w := &Syncer{
+		logger:         logger,
+		config:         config,
+		writer:         writer,
+		syncer:         syncer,
+		timer:          clock.Timer(config.IntervalTrigger),
+		wg:             &sync.WaitGroup{},
+		bytesToSync:    atomic.NewUint64(0),
+		bytesTriggerCh: make(chan struct{}, 1),
+		syncLock:       &sync.Mutex{},
+		notifierLock:   &sync.RWMutex{},
+		notifier:       notify.New(),
+		opFn:           opFn,
+	}
+
+	w.ctx, w.cancel = context.WithCancel(ctx)
+
+	w.logger.Infof(`sync is enabled, mode=%s, sync each "%s" or "%s"`, config.Mode, config.IntervalTrigger, config.BytesTrigger)
+
+	w.startSyncLoop()
+
+	return w
+}
+
+// Write writes to the underlying writer and trigger synchronization when the configured Config.BytesTrigger volume is exceeded.
+func (s *Syncer) Write(p []byte) (n int, err error) {
+	if err := s.ctx.Err(); err != nil {
+		return 0, errors.Errorf(`syncer is closed: %w`, err)
+	}
+
+	// Write data to the underlying writer
+	n, err = s.writer.Write(p)
+	if err != nil {
+		return n, err
+	}
+
+	s.onWrite(n)
+	return n, nil
+}
+
+// WriteString - see Write,
+// some writes are optimized for writing strings
+// without an unnecessary conversion from []byte to string,
+// so both methods are supported.
+func (s *Syncer) WriteString(str string) (n int, err error) {
+	if err := s.ctx.Err(); err != nil {
+		return 0, errors.Errorf(`syncer is closed: %w`, err)
+	}
+
+	// Write data to the underlying writer
+	n, err = s.writer.WriteString(str)
+	if err != nil {
+		return n, err
+	}
+
+	s.onWrite(n)
+	return n, nil
+}
+
+// Close method stop periodical synchronization.
+func (s *Syncer) Close() error {
+	s.logger.Debug(`closing syncer`)
+	s.cancel()
+	s.wg.Wait()
+	s.SyncAndWait()
+	s.logger.Debug(`syncer closed`)
+	return nil
+}
+
+// Notifier returns Notifier that will inform about the next sync.
+// If synchronization is disabled, or waiting is disabled (Config.Wait==false),
+// the *notify.Notifier(nil) value is returned,
+// so the Notifier.Wait() method returns immediately.
+func (s *Syncer) Notifier() *notify.Notifier {
+	if s == nil || !s.config.Wait {
+		// Note: *notify.Notifier(nil).Wait() is a valid call
+		return nil
+	}
+
+	s.notifierLock.RLock()
+	notifier := s.notifier
+	s.notifierLock.RUnlock()
+
+	return notifier
+}
+
+// DoWithNotifier guarantees that the Notifier will not change during the DO operation, the sync will not start.
+// After the DO operation, notifier is returned, so you can wait for the next sync.
+// If synchronization is disabled, or waiting is disabled (Config.Wait==false),
+// the *notify.Notifier(nil) value is returned,
+// so the Notifier.Wait() method returns immediately.
+func (s *Syncer) DoWithNotifier(do func()) *notify.Notifier {
+	if s == nil || !s.config.Wait {
+		do()
+		// Note: *notify.Notifier(nil).Wait() is a valid call
+		return nil
+	}
+
+	s.notifierLock.RLock()
+	notifier := s.notifier
+	do()
+	s.notifierLock.RUnlock()
+
+	return notifier
+}
```

Comment:
Jaký má důvod, že tomu předáváš callback? Přijde mě to zvláštní, protože v tetech kromě assertu tam voláš jen Write|WriteString, potřebuješ pak vůbec Write|WriteString? Třeba správně nechápu ten use-case a v testu není vidět.

Reply:
@zajca Ide o zamok `RLock`/ `RUnlock`. 

Samozrejme, mohli by sa namiesto `Syncer.DoWithNotifier` spravit metody `Syncer.RLock/RUnlock` a callback by tam nebol. Riesilo by sa to o uroven vyssie. Toto mi prislo ako viac odolne riesenie, neda sa zabudnut odomknut zamok (v testoch to je pokryte).

Teda tym callbackom sa hovori, "pocas tejto operacie sa nesmie vymenit notifier".

Vyvetlim na priklade.

**1000x zapisov caka na sync na disk**
  - Pretoze klientovi povime `OK`, az ked su data fyzicky na disku
  - Cakanie riesi `notify.Notifier.Wait()`, tu sa necaka, len sa vrati notifier.
  - Teda `DoWithNotifier` prebehne 1000x a vrati vzdy referenciu na rovnaky `Notifier`.

----------------------------
     
**No a potom mame `doSync`, v tomto priklade povedzme, ze volany kazdych `100ms`.**
https://github.com/keboola/keboola-as-code/blob/23aa88c6d8ba05732988772b7a915c2e4662eecc/internal/pkg/service/buffer/storage/local/writer/disksync/disksync.go#L239-L244
- Tu uz nie je `RLock/RUnlock` ale `Lock/Unlock`.
- Read zamkov mozes mat viacero, Write zamkov iba jeden naraz.
- `Lock` pocka na tych 1000x beziach zapisov (jeden zapis su mikrosekundy).
- Ked pride 1001 zapis, tak ten je uz zablokovany, tak funguje `*sync.RWLock`, je to ako keby fronta, kde ma zapis prednost pred citanim (ale beziace citanie musi najprv skoncit).
  - Viac info: `a blocked Lock call excludes new readers from acquiring the lock.` https://golang.org/pkg/sync/#RWMutex

- Teda:
```
1 RLock
....
1000 RLock
Lock
1001 RLock
```
- Kazdy zapis uz prebehol, sme v situacii, kde 1000 write operacii ma referenciu na `NOTIFIER1` a caka `Notifier.Wait()`.
- `Lock` v `doSync` je uz odblokovany, vymenime teda `notifier` za novy `NOTIFIER2`, a ulozime si referenciu na stary `NOTIFIER1`.
   - Vymena je otazka par mikrosekund.
   - Spustime synchronizaciu.
   - Zavolame `NOTIFIER1.Done()`, lebo vsetky write operacie, ktore prebehli pree vymenou notifiers su uz fyzicky na disku.
https://github.com/keboola/keboola-as-code/blob/23aa88c6d8ba05732988772b7a915c2e4662eecc/internal/pkg/service/buffer/storage/local/writer/disksync/disksync.go#L272-L273
   - Kym robime sync, tak bezia dalsie write operacie, ktore ale dostavaju uz `NOTIFIER2`.

--------------------

**Co by sa stalo bez zamkov?**

Atualny kod:
```
// Write
notifier := syncer.DoWithNotifier() {
    writer.WriteString("foo")
}

// Wait for sync and return sync error, if any
return notifier.Wait()
```

----------------

**Bez zamku variant 1.**

```
notifier := sync.Notifier()
<<< sync triggered/done, a new notifier is created >>>
writer.WriteString("foo")
return notifier.Wait()
```

V tomto pripade by sme vratili klientovi, ze data su bezpecne na disku, aj ked na disku byt nemusia.
 
 ----------------

**Bez zamku variant 2.**

```

writer.WriteString("foo")
<<< sync triggered/done, a new notifier is created>>>
return sync.Notifier().Wait()
```

A tu je to naopak, data su uz na disku, ale my mame notifier, ktory odblokuje az dalsia synchronizacia.
To nerobi problem pre klienta, len sa v najhorsom pripade caka 2x tak dlho.

^^^^ Takto som to mal este pred par dnami, ale je to blbo testovatelne. Spustil som testy napr. 10000x, a 2x sa stalo ze mi medzi `Write` a `Wait` vosiel sync. V produkcii nevadi, ale test to kvalitne zablokovalo, kedze ten dalsi sync tam napisany uz nie je.

---------

Bez callbaku by to bolo:
```
syncer.RLock()
notifier := syncer.Notifier()
writer.WriteString("foo")
syncer.RUnlock()
return notifier.Wait()
```

^^^ a to som nechcel riesi/komplikoat v nadradenom kode.

-----------------------


 Da sa tomu ^^^ nejak pochopit? :slightly_smiling_face: 
 
 Mozno by sa lepsie citalo nejake riesenie zlozene na `timestampoch`, ze si ulozime cas startu `Sync` a vsetky writes pred tymto casom potom odblokujeme. Alebo si rovno ulozime cakajuce write operacie do nejakeho poľa. 
 
 Zial, tieto ^^^ ine alternativy su pomalsie a vyzaduju si nejaku alokaciu pamate. Aktualne riesenie je postavene iba na zamkoch a jednom channel v `notify.Notifier`, ktory ale sluzi pre 1000cky zapisov.
 
 ---------------
 
 Mozeme si zavolat ak to takto nestaci. 
 Kludne navrhni nejake ine riesenie, alebo co presne nebolo jasne a zdokumentujem to lepsie.

Reply:
Asi to nemá smysl nijak blokovat, chápu co tam řeší ten zákem, moc sa mě nelíbý to api kdy přédpokládáš, že v tom callbacku něco uděláš.
něco jako DoWriteWithNatifier DoWriteStringWithNotifier by to mohly řešit, ale je to asi zbytečné, možná je to můj PHP mindrák, že používaní callables to celé hlavně zkomplikuješ :D